oper8/deploy_manager/replace_utils.py

```diff
@@ -0,0 +1,109 @@
+        for i in range(min(len(current_manifest), len(desired_manifest))):
+            if modified_lists(current_manifest[i], desired_manifest[i], in_list=True):
+                return True
+    if isinstance(current_manifest, dict) and isinstance(desired_manifest, dict):
+        key_intersection = set(current_manifest.keys()).intersection(
+            set(desired_manifest.keys())
+        )
+        # If there are no common keys and we're in a list then return True
+        # this means we have a new object
+        if in_list and not key_intersection:
+            log.debug4("Requires replace due to no common key in list")
+            return True
+
+        # Tack if one key has the same value. This is
+        # required for kubernetes merges
+        at_least_one_common = False
+        for k in key_intersection:
+            # Check if two objects are the same for their value operations
+            changed = False
+            if isinstance(current_manifest[k], list):
+                changed = bool(
+                    recursive_list_diff(current_manifest[k], desired_manifest[k])
+                )
+            elif isinstance(current_manifest[k], dict):
+                changed = bool(recursive_diff(current_manifest[k], desired_manifest[k]))
+            else:
+                changed = current_manifest[k] != desired_manifest[k]
+
+            if not changed:
+                at_least_one_common = True
+            if modified_lists(current_manifest[k], desired_manifest[k]):
+                return True
+        if in_list and not at_least_one_common:
+            log.debug4("Requires replace due to no common key/value in list")
+            return True
+    return False
+
+
+def modified_value_from(current_manifest: Any, desired_manifest: Any) -> bool:
+    """Helper function to check if a manifest switched from value to valueFrom. These are mutually
+    exclusive thus they require a replace command.
+    """
+    if isinstance(current_manifest, list) and isinstance(desired_manifest, list):
+        iteration_len = min(len(current_manifest), len(desired_manifest))
+        for i in range(iteration_len):
+            if modified_value_from(current_manifest[i], desired_manifest[i]):
+                return True
+    if isinstance(current_manifest, dict) and isinstance(desired_manifest, dict):
+        if ("value" in current_manifest and "valueFrom" in desired_manifest) or (
+            "valueFrom" in current_manifest and "value" in desired_manifest
+        ):
+            log.debug4("Requires replace due to value/valueFrom change")
+            return True
+        else:
+            for k in set(current_manifest.keys()).intersection(
+                set(desired_manifest.keys())
+            ):
+                if modified_value_from(current_manifest[k], desired_manifest[k]):
+                    return True
+    return False
+
+
+REPLACE_FUNCS: List[Callable[[str, str], bool]] = [modified_lists, modified_value_from]
```

Comment:
Is this here because it's needed elsewhere? Seems like this whole module could boil down to expose only `requires_replace` as its public API and the rest could be implementation details that are only used here and in unit tests

Reply:
It could, I included the list incase it needed to be modified quickly by a consumer. It can be made private though